server/src/main/java/org/elasticsearch/plugins/SearchPlugin.java

```diff
@@ -124,6 +127,15 @@
         return emptyList();
     }
 
+    /**
+     * Allows plugins to register cache differentiators which contributes to the cacheKey
+     * computation for the request cache. This helps differentiate between queries that
+     * are otherwise identical.
+     */
+    default List<CheckedBiConsumer<ShardSearchRequest, StreamOutput, IOException>> getRequestCacheKeyDifferentiators() {
+        return emptyList();
+    }
```

Comment:
+1, agreed, we already limit the number of reader wrapper to 1 so that makes sense to have this limitation here too.
We don't expect usage outside of the security use case so no need to add this flexibility upfront.

Reply:
Thanks both. I have [pushed](https://github.com/elastic/elasticsearch/pull/70191/commits/b576759c2ad7610a3fe56d53ddc9795b12917017) the following changes based on the suggestions:
* The extension method now returns a `RequestCacheKeyProvider` instead of a differentiator to better encapsulate the cache key building logic
* Only allow a single `RequestCacheKeyProvider` to be provided by `SearchPlugin`(s). If none is provided, it by defaults uses the current implementation from `ShardSearchRequest`.